server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -426,46 +425,75 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId, Action
                 logger.warn(() -> new ParameterizedMessage("cannot read snapshot file [{}]", snapshotId), ex);
             }
             // Delete snapshot from the index file, since it is the maintainer of truth of active snapshots
-            final RepositoryData repositoryData;
             final RepositoryData updatedRepositoryData;
+            final BlobContainer indicesBlobContainer = blobStore().blobContainer(basePath().add("indices"));
+            final Map<String, BlobContainer> foundIndices;
             try {
-                repositoryData = getRepositoryData();
+                final RepositoryData repositoryData = getRepositoryData();
                 updatedRepositoryData = repositoryData.removeSnapshot(snapshotId);
+                // Cache the indices that were found before writing out the new index-N blob so that a stuck master will never
```

Comment:
I think this approach of listing before writing the updated `index-N` makes this approach completely safe even on an eventually consistent blob store. If an index uuid (the snapshot uuid, not the CS one!) goes out of scope here for a single `index-N` it will never be reused in a future `index-M` (`M > N)` => we don't have to care about stuck master nodes coming back to haunt us.
What we could however do is optimize this a little and maybe write a `SUCCESS-N` blob after we finished all operations that went into a given `index-N` blob and only run the listing if we suspect dangling indices because an operation that wrote a given `index-N` didn't finish fully.
The downside of this is, that it gets a little complicated when you have to account for master failover scenarios and may not be worth it performance wise now that we have bulk deletes and could parallelize the recursive delete as well.